src/EFCore.PG/Storage/Internal/Mapping/NpgsqlRangeTypeMapping.cs

```diff
@@ -50,20 +50,32 @@ protected NpgsqlRangeTypeMapping(RelationalTypeMappingParameters parameters, Npg
             : base(parameters, npgsqlDbType) { }
 
         public override RelationalTypeMapping Clone(string storeType, int? size)
-            => new NpgsqlRangeTypeMapping<T>(Parameters.WithStoreTypeAndSize(storeType, size), NpgsqlDbType);
+            => new NpgsqlRangeTypeMapping(Parameters.WithStoreTypeAndSize(storeType, size), NpgsqlDbType);
 
         public override CoreTypeMapping Clone(ValueConverter converter)
-            => new NpgsqlRangeTypeMapping<T>(Parameters.WithComposedConverter(converter), NpgsqlDbType);
+            => new NpgsqlRangeTypeMapping(Parameters.WithComposedConverter(converter), NpgsqlDbType);
 
         protected override string GenerateNonNullSqlLiteral(object value)
         {
-            var range = (NpgsqlRange<T>)value;
             var sb = new StringBuilder();
             sb.Append('\'');
-            sb.Append(range.ToString());
+            sb.Append(value);
             sb.Append("'::");
             sb.Append(StoreType);
             return sb.ToString();
         }
+
+        static NpgsqlDbType GenerateNpgsqlDbType(RelationalTypeMapping subtypeMapping)
+        {
+            if (subtypeMapping is NpgsqlTypeMapping npgsqlTypeMapping)
+                return NpgsqlDbType.Range | npgsqlTypeMapping.NpgsqlDbType;
+
+            // We're using a built-in, non-Npgsql mapping such as IntTypeMapping.
+            // Infer the NpgsqlDbType from the DbType (somewhat hacky but why not).
+            Debug.Assert(subtypeMapping.DbType.HasValue);
```

Comment:
Couldn't we just use `subtypeMapping.DbType ?? throw ...`?

Reply:
Well, `subtypeMapping.DbType` should never be null - this is a programming assertion rather than a runtime check (which would imply this scenario is possible but not supported in some way)... It's a good idea to use assertions for invariants rather than exceptions, it documents the code better.